source/base/PrimaryGeneration.h

```diff
@@ -29,21 +28,21 @@ namespace nexus {
     void GeneratePrimaries(G4Event*);
 
     /// Sets the primary generator
-    void SetGenerator(G4VPrimaryGenerator*);
+    void SetGenerator(std::unique_ptr<G4VPrimaryGenerator> pg);
     /// Returns a pointer to the primary generator
     const G4VPrimaryGenerator* GetGenerator() const;
 
   private:
-    G4VPrimaryGenerator* generator_; ///< Pointer to the primary generator
+    std::unique_ptr<G4VPrimaryGenerator> generator_; ///< Pointer to the primary generator
   };
 
   // INLINE DEFINITIONS //////////////////////////////////////////////
 
-  inline void PrimaryGeneration::SetGenerator(G4VPrimaryGenerator* pg)
-  { generator_ = pg; }
+  inline void PrimaryGeneration::SetGenerator(std::unique_ptr<G4VPrimaryGenerator> pg)
```

Comment:
I'd do here the initialisation of the `std::unique_ptr` that you are doing right now in the class `NexusApp`.